curl-helper.c

```diff
@@ -1250,25 +1251,102 @@ static int cb_OPENSOCKETFUNCTION(void *data,
     CAMLparam0();
     CAMLlocal1(result);
     Connection *conn = (Connection *)data;
-    int sock = -1;
+    curl_socket_t sock = CURL_SOCKET_BAD;
     (void)purpose; /* not used */
 
     sock = socket(addr->family, addr->socktype, addr->protocol);
 
-    if (-1 != sock)
+    if (CURL_SOCKET_BAD != sock)
     {
       /* FIXME windows */
       result = caml_callback_exn(Field(conn->ocamlValues, Ocaml_OPENSOCKETFUNCTION), Val_int(sock));
       if (Is_exception_result(result))
       {
         close(sock);
-        sock = -1;
+        sock = CURL_SOCKET_BAD;
       }
     }
     CAMLdrop;
 
     caml_release_runtime_system();
-    return ((sock == -1) ? CURL_SOCKET_BAD : sock);
+    return sock;
+}
+
+static curl_socket_t cb_OPENSOCKETFUNCTION2(void *data,
+                        curlsocktype purpose,
+                        struct curl_sockaddr *addr)
+{
+    caml_acquire_runtime_system();
+
+    CAMLparam0();
+    CAMLlocal5(result, v_sock_purpose, v_sockdomain, v_socktype, v_sockaddr_record);
+    Connection *conn = (Connection *)data;
+    curl_socket_t sock = CURL_SOCKET_BAD;
+
+    switch (purpose) {
+        case CURLSOCKTYPE_IPCXN:
+            v_sock_purpose = Val_int(0);
+            break;
+        default:
+            goto cleanup;
+    }
+
+    switch (addr->family) {
+        case AF_UNIX:
+            v_sockdomain = Val_int(0); /* PF_UNIX */
+            break;
+        case AF_INET:
+            v_sockdomain = Val_int(1); /* PF_INET */
+            break;
+        case AF_INET6:
+            v_sockdomain = Val_int(2); /* PF_INET6 */
+            break;
+        default:
+            goto cleanup;
+    }
+
+    switch (addr->socktype) {
+        case SOCK_STREAM:
+            v_socktype = Val_int(0); /* SOCK_STREAM */
+            break;
+        case SOCK_DGRAM:
+            v_socktype = Val_int(1); /* SOCK_DGRAM */
+            break;
+        case SOCK_RAW:
+            v_socktype = Val_int(2); /* SOCK_RAW */
+            break;
+        case SOCK_SEQPACKET:
+            v_socktype = Val_int(3); /* SOCK_SEQPACKET */
+            break;
+        default:
+            goto cleanup;
+    }
+
+    union sock_addr_union sock_addr;
+    socklen_t addrlen = addr->addrlen;
+    if (addrlen > sizeof(sock_addr)) addrlen = sizeof(sock_addr);
```

Comment:
how is this possible and if it gets truncated like that then result is meaningless?

Reply:
same as now, so goto